Add docs for realtime decoding (#345)
NVIDIA/cudaqx: docs/sphinx/examples_rst/qec/realtime_decoding.rst

- Syndrome extraction: Measures the stabilizers of the logical qubits.

- Correction application: Applies the corrections to the logical qubits.

- Logical observable measurement: Measures the logical observables of the logical qubits.

- Decoder finalization: Frees up resources after circuit execution.

The API is designed to be called from within quantum kernels (marked with ``@cudaq.kernel`` in Python or ``__qpu__``  in C++). The runtime automatically routes these calls to the appropriate backend—whether a simulation environment on your local machine or a low-latency connection to quantum hardware. The API is device-agnostic, so the same kernel code works across different deployment scenarios.

However, the user is still required to provide a configuration file, or generate one if it is not present. The generation process depends on the decoder type and the detector error model studied in other sections of the documentation. Moreover, the user must write an appropriate kernel that describes the correct syndrome extraction and correction application logic.

> Please, follow the instructions in the next section to generate a configuration file, write a quantum kernel, and compile and run the examples correctly.

Reviewer: ```suggestion
Follow the instructions in the next section to generate a configuration file, write a quantum kernel, compile and run the examples.
```

The original reads a bit harsh to me

Configuration
-------------

> The configuration process transforms your quantum circuit's error characteristics into a format that decoders can efficiently process. This section walks through each step in detail, showing how to go from circuit simulation to a fully configured real-time decoder.

Reviewer 2: Looks like this was updated in b2e47f9583dcfb56d7e8f6730b2b753741e2613c

      multi_config.decoders = [config]
      
      yaml_str = multi_config.to_yaml_str(200)
      with open("decoder_config.yaml", "w") as f:
          f.write(yaml_str)
      
      # Load configuration
      status = qec.configure_decoders_from_file("decoder_config.yaml")

.. tab:: C++

   .. code-block:: cpp

      multi_decoder_config multi_config;
      multi_config.decoders.push_back(config);
      
      std::string yaml_str = multi_config.to_yaml_str(200);
      std::ofstream file("decoder_config.yaml");
      file << yaml_str;
      file.close();
      
      // Load configuration
      int status = configure_decoders_from_file("decoder_config.yaml");

Step 4: Use in Quantum Kernels
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

With your decoders configured and initialized, you're ready to use them within quantum kernels. The real-time decoding API provides three key functions that integrate seamlessly with CUDA-Q's quantum programming model: ``reset_decoder`` prepares a decoder for a new shot, ``enqueue_syndromes`` sends syndrome measurements to the decoder for processing, and ``get_corrections`` retrieves the decoder's recommended corrections.

> These functions are designed to be called from within quantum kernels (marked with ``@cudaq.kernel`` in Python or ``__qpu__`` in C++). The runtime automatically routes these calls to the appropriate backend - whether that's a simulation environment on your local machine or a low-latency connection to quantum hardware. The API is device-agnostic, so the same kernel code works across different deployment scenarios.

Reviewer 2: I'm not sure what it looked like when you commented @melody-ren, but here is what they are now.

```
$ grep -n -P "^These functions" ../docs/sphinx/examples_rst/qec/realtime_decoding.rst
94:These functions take a quantum code, an initial state preparation operation, the number of measurement rounds, and a noise model, then return a complete detector error model ready for decoder configuration. The user simply needs to configure the noise model and specify the circuit structure—the library handles all the error tracking and matrix construction automatically.
192:These functions are designed to be called from within quantum kernels (marked with ``@cudaq.kernel`` in Python or ``__qpu__`` in C++). The runtime automatically routes these calls to the appropriate backend - whether that is a simulation environment on the local machine or a low-latency connection to quantum hardware. The API is device-agnostic, so the same kernel code works across different deployment scenarios.
```

They start of the same but look a little different now. @melody-ren please mark this thread as resolve if you think it is resolved now.

          
          # Measure logical observable
          result = mz(data)

.. tab:: C++

   .. code-block:: cpp

      #include "cudaq/qec/realtime/decoding.h"
      
      __qpu__ void qec_circuit(int num_rounds, int decoder_id) {
          // Reset decoder
          cudaq::qec::decoding::reset_decoder(decoder_id);
          
          // Allocate qubits
          cudaq::qvector data(25), ancx(12), ancz(12);
          cudaq::qec::patch logical(data, ancx, ancz);
          
          // Prepare logical state
          prep0(logical);
          
          // Syndrome extraction with real-time decoding
          for (int round = 0; round < num_rounds; ++round) {
              auto syndromes = measure_stabilizers(logical);
              cudaq::qec::decoding::enqueue_syndromes(decoder_id, syndromes);
          }
          
          // Get and apply corrections
          auto corrections = cudaq::qec::decoding::get_corrections(
              decoder_id, 1, false);
          if (corrections[0]) {
              cudaq::x(data);  // Apply correction
          }
          
          // Measure logical observable
          auto result = mz(data);
      }

> Decoder Types

Author: I agree with you but I would make it an action item for after Monday.
We have to write actually lots of doxygens :)

> Decoder Types

Reviewer 2: > I agree with you but I would make it an action item for after Monday. We have to write actually lots of doxygens :)

https://github.com/NVIDIA/cudaqx/issues/358 written.

The decoder offers extensive tunability. You can adjust the number of BP iterations to trade off latency for accuracy, choose between sum-product and min-sum BP variants, and control OSD search depth. For real-time applications, start with conservative settings (50 iterations, OSD order 7) and tune based on your observed error rates and latency requirements.

* **Best for**: Medium to large codes (distance ≥ 7), moderate to high error rates, scenarios where GPU acceleration is available
* **Key Parameters**:
  
  * ``error_rate_vec`` (list/vector of floats): Per-mechanism error probabilities - crucial for BP convergence. These should match your DEM's error rates.
  * ``max_iterations`` (int): Maximum BP iterations (typically 50-100). More iterations improve accuracy but increase latency.

>   * ``use_osd`` (bool): Enable Ordered Statistics Decoding post-processing when BP fails to converge. Recommended for production use.

Reviewer: Seems like abd874b8ef52d2840446686c0c87dacbc07d4362 fixed this

  * ``osd_order`` (int): OSD search depth (typically 7-10). Higher values are more thorough but slower.
  * ``bp_method`` (int): BP algorithm variant (0: sum-product, 1: min-sum). Min-sum is faster but slightly less accurate.

.. tab:: Python

   .. code-block:: python

      nv_config = qec.nv_qldpc_decoder_config()
      nv_config.error_rate_vec = [0.01] * config.block_size
      nv_config.max_iterations = 50
      nv_config.use_osd = True
      nv_config.osd_order = 7
      config.set_decoder_custom_args(nv_config)

.. tab:: C++

   .. code-block:: cpp

      nv_qldpc_decoder_config nv_config;
      nv_config.error_rate_vec = error_rates;
      nv_config.max_iterations = 50;
      nv_config.use_osd = true;
      nv_config.osd_order = 7;
      config.decoder_custom_args = nv_config;

sliding_window
^^^^^^^^^^^^^^
The windowed decoder is designed for larger codes where full maximum-likelihood decoding is impractical. 
It works by dividing the decoding problem into overlapping "windows" of syndrome data, 

> decoding each window independently (using a standard inner decoder), 

Reviewer: ```suggestion
decoding each window independently (using an inner decoder), 
```

and then combining the results to form a global correction. 
This approach reduces memory and computational requirements while still capturing most local error correlations.

* **Best for**: Very long circuits, memory-constrained systems
* **Parameters**:
  
  * ``window_size``: Number of rounds per window
  * ``step_size``: Window advancement (equals window_size for non-overlapping)
  * ``num_syndromes_per_round``: Syndromes per round

>   * ``inner_decoder_name``: Type of decoder for each window

Reviewer: ```suggestion
  * ``inner_decoder_name``: Type of decoder for the windows
```

"each window" reads like each window can have its own inner decoder

The files will have names like ``surface_code-1.cpp`` and ``surface_code_1.py``.

We hope that the user felt supported in the development of their application, and that enough information was provided to guide them in the process.
The subsequent step, once the user has chosen the appropiate decoder and the appropiate backend, is to compile and execute the application.
We will provide the instructions to do so for both the simulation and the hardware backends.

C++ Compilation
^^^^^^^^^^^^^^^

**Simulation Backend (Stim)**

Compile with the simulation backend for local testing:

.. code-block:: bash

   nvq++ --target stim my_circuit.cpp \
         -lcudaq-qec \
         -lcudaq-qec-realtime-decoding \
         -lcudaq-qec-realtime-decoding-simulation \
         -o my_circuit_sim

   # Execute
   ./my_circuit_sim --distance 3 --num_shots 1000 --save_dem config.yaml

**Key Points:**

- ``--target stim``: Use the Stim quantum simulator
- ``-lcudaq-qec``: Core QEC library with codes and experiments
- ``-lcudaq-qec-realtime-decoding``: Real-time decoding core API
- ``-lcudaq-qec-realtime-decoding-simulation``: Simulation-specific decoder backend

**Quantinuum Backend (Emulation)**

> Compile for Quantinuum emulation mode:

Reviewer 2: When running with `--emulate`, there is no noise being applied because there is no way to express noise in platform-agnostic QIR. Therefore, when running with emulation, users will see noise-free sample data. That should probably be mentioned here.

Reviewer 2: Updated in 80ee1b5

         -Wl,--export-dynamic \
         -o my_circuit_quantinuum

   # Execute
   ./my_circuit_quantinuum --distance 3 --num_shots 1000 --load_dem config.yaml

**Key Points:**

- ``--target quantinuum --emulate``: Use Quantinuum emulator
- ``--quantinuum-machine Helios-Fake``: Specify machine (``Helios-Fake`` for emulation)
- ``-lcudaq-qec-realtime-decoding-quantinuum``: Quantinuum-specific decoder backend (replaces ``-simulation``)
- ``-Wl,--export-dynamic``: **Required** linker flag for dynamic symbol resolution

**Quantinuum Backend (Hardware)**

Compile for actual Quantinuum hardware:

.. code-block:: bash

   nvq++ --target quantinuum \
         --quantinuum-machine H2-1 \
         my_circuit.cpp \
         -lcudaq-qec \
         -lcudaq-qec-realtime-decoding \
         -lcudaq-qec-realtime-decoding-quantinuum \
         -Wl,--export-dynamic \
         -o my_circuit_hardware

   # Execute
   export CUDAQ_QUANTINUUM_CREDENTIALS=<your_credentials_file>
   ./my_circuit_hardware --distance 3 --num_shots 100 --load_dem config.yaml

**Key Points:**

- Remove ``--emulate`` flag for hardware execution
- Use real machine names: ``H2-1``, ``H2-2``, etc.

> - Set ``CUDAQ_QUANTINUUM_CREDENTIALS`` environment variable with your credentials

Reviewer 2: Can you please add a link to https://nvidia.github.io/cuda-quantum/latest/using/backends/hardware/iontrap.html#quantinuum so that users can see where to get more detailed information about running on Quantinuum hardware with CUDA-Q in general?

      config.block_size = 10
      config.syndrome_size = 5
      config.H_sparse = [0, 1, -1, 1, 2, -1]  # Minimal test data
      config.O_sparse = [0, -1]
      config.D_sparse = [0, -1]
      
      lut_config = qec.multi_error_lut_config()
      lut_config.lut_error_depth = 1
      config.set_decoder_custom_args(lut_config)
      
      multi_config = qec.multi_decoder_config()
      multi_config.decoders = [config]
      
      status = qec.configure_decoders(multi_config)
      print(f"Configuration status: {status}")
      
      qec.finalize_decoders()
      print("Setup verified!")

.. tab:: C++

   .. code-block:: bash

      # Compile test
      nvq++ --target stim test_setup.cpp \
            -lcudaq-qec \
            -lcudaq-qec-realtime-decoding \
            -lcudaq-qec-realtime-decoding-simulation
      
      # Run
      ./a.out

If the test completes without errors, your setup is ready for real-time decoding experiments.

Best Practices
--------------

Successfully deploying real-time decoding requires attention to several key details. These best practices emerge from real-world usage and help you avoid common pitfalls while optimizing performance.

> Decoder Selection

Reviewer 2: It looks like this was addressed in b2e47f9583dcfb56d7e8f6730b2b753741e2613c

      sw_config.num_syndromes_per_round = 24;
      sw_config.inner_decoder_name = "multi_error_lut";
      multi_error_lut_config inner_config;
      inner_config.lut_error_depth = 2;
      sw_config.multi_error_lut_params = inner_config;
      config.decoder_custom_args = sw_config;

Backend Selection
-----------------

CUDA-Q QEC's real-time decoding system is designed to work seamlessly across different execution environments. The backend selection determines where quantum circuits run and how decoders communicate with the quantum processor. Understanding the differences between simulation and hardware backends helps the user develop efficiently and deploy confidently.

Simulation Backend
^^^^^^^^^^^^^^^^^^

The simulation backend is the primary tool during development, testing, and algorithm validation. It runs entirely on the local machine, using quantum simulators like Stim to execute circuits while decoders process syndromes in parallel threads. This setup is ideal for rapid iteration: the user can test decoder configurations, validate circuit logic, and debug syndrome processing without waiting for hardware access or paying for compute time.

> The simulation backend faithfully mimics real-time decoding's concurrent operation - decoders run in separate threads and process syndromes asynchronously, just as they would on hardware. This means code will behave the same way whether testing locally or running on a quantum computer. The main difference is that simulation does not have the same strict latency constraints, making it easier to experiment with complex decoder configurations.

Reviewer 2: The note about separate threads is only true for special configurations that we were unable to make public. Our "regular" simulation backend runs in the same thread as the CUDA-Q kernel itself.

Reviewer 2: Fixed in ee6cd7b

Use the simulation backend for local development and testing:

.. tab:: Python

   .. code-block:: python

      import cudaq
      import cudaq_qec as qec
      
      cudaq.set_target("stim")  # Or other simulator
      qec.configure_decoders_from_file("config.yaml")
      
      # Run circuit with noise model
      results = cudaq.run(my_circuit, shots_count=100, 
                         noise_model=cudaq.NoiseModel())

.. tab:: C++

   .. code-block:: bash

      # Compile with simulation support
      nvq++ -std=c++20 my_circuit.cpp -lcudaq-qec \

>             -lcudaq-qec-realtime-simulation

Reviewer 2: Fixed in ee6cd7b

A full set of common examples is provided to guide development.
These examples describe the complete workflow for developing an application that uses real-time decoding in a single file.
The relevant examples can be found at the following path: ``libs/qec/unittests/realtime/app_examples/`` (C++) and the corresponding Python examples.
The files have names like ``surface_code-1.cpp`` and ``surface_code_1.py``.

These examples provide comprehensive support for application development with real-time decoding.
The subsequent step, once the user has chosen the appropriate decoder and the appropriate backend, is to compile and execute the application.
Instructions are provided below for both the simulation and the hardware backends.

C++ Compilation
^^^^^^^^^^^^^^^

**Simulation Backend (Stim)**

Compile with the simulation backend for local testing:

.. code-block:: bash

   nvq++ --target stim my_circuit.cpp \
         -lcudaq-qec \
         -lcudaq-qec-realtime-decoding \
         -lcudaq-qec-realtime-decoding-simulation \
         -o my_circuit_sim

   # Execute

>    ./my_circuit_sim --distance 3 --num_shots 1000 --save_dem config.yaml

Reviewer 2: The examples on this page do not support these command-line arguments.

**Key Points:**

- ``--target stim``: Use the Stim quantum simulator
- ``-lcudaq-qec``: Core QEC library with codes and experiments
- ``-lcudaq-qec-realtime-decoding``: Real-time decoding core API
- ``-lcudaq-qec-realtime-decoding-simulation``: Simulation-specific decoder backend

**Quantinuum Backend (Emulation)**

Compile for Quantinuum emulation mode:

.. code-block:: bash

   nvq++ --target quantinuum --emulate \
         --quantinuum-machine Helios-Fake \
         my_circuit.cpp \
         -lcudaq-qec \
         -lcudaq-qec-realtime-decoding \
         -lcudaq-qec-realtime-decoding-quantinuum \
         -Wl,--export-dynamic \
         -o my_circuit_quantinuum

   # Execute

>    ./my_circuit_quantinuum --distance 3 --num_shots 1000 --load_dem config.yaml

Reviewer 2: The examples on this page do not support these command-line arguments.

**Key Points:**

- ``--target quantinuum --emulate``: Use Quantinuum emulator
- ``--quantinuum-machine Helios-Fake``: Specify machine (``Helios-Fake`` for emulation)
- ``-lcudaq-qec-realtime-decoding-quantinuum``: Quantinuum-specific decoder backend (replaces ``-simulation``)
- ``-Wl,--export-dynamic``: **Required** linker flag for dynamic symbol resolution

**Quantinuum Backend (Hardware)**

Compile for actual Quantinuum hardware:

.. code-block:: bash

   nvq++ --target quantinuum \
         --quantinuum-machine Helios-1 \
         my_circuit.cpp \
         -lcudaq-qec \
         -lcudaq-qec-realtime-decoding \
         -lcudaq-qec-realtime-decoding-quantinuum \
         -Wl,--export-dynamic \
         -o my_circuit_hardware

   # Execute
   export CUDAQ_QUANTINUUM_CREDENTIALS=<credentials_file_path>

>    ./my_circuit_hardware --distance 3 --num_shots 100 --load_dem config.yaml

Reviewer 2: The examples on this page do not support these command-line arguments.

   # Configure target for hardware
   cudaq.set_target("quantinuum",
                    emulate=False,       # Hardware execution
                    machine="Helios-1",  # Use actual hardware
                    extra_payload_provider="decoder")
   
   # Load decoder configuration
   qec.configure_decoders_from_file("config.yaml")
   
   # Run circuit
   result = cudaq.run(my_circuit, shots_count=100)  # Fewer shots for hardware
   
   # Cleanup
   qec.finalize_decoders()

**Key Points:**

- ``emulate=False``: Execute on real quantum hardware
- Use real machine names (check Quantinuum portal for available machines)
- Reduce shot count for hardware experiments (hardware time is expensive)

Complete Workflow Example
^^^^^^^^^^^^^^^^^^^^^^^^^^
Given that the user follows the structure of the examples provided, where each executable takes terminal arguments to configure the application, the following workflow can be used to compile and execute the application.

.. code-block:: bash

   # Phase 1: Generate Detector Error Model (DEM)
   # This is done once per code/distance/noise configuration
   
   ## C++
   ./my_circuit_sim --distance 3 --num_shots 1000 --p_spam 0.01 \

>                     --save_dem config_d3.yaml --num_rounds 12 --decoder_window 6

Reviewer 2: The examples in this page do not accept the command-line arguments being shown here.

This comment applies to many of the commands below.

Reviewer 2: I see now that this one section is referring to the surface code examples in the GitHub repo, not the example on this webpage. Correspondingly, I took a pass at making some updates to make this more clear: f9535fd

4. **Configuration Lifetime**: Decoder configurations persist on Quantinuum's servers and are referenced by UUID. If the configuration is modified, it must be uploaded again - the system will generate a new UUID and use the new configuration for subsequent jobs.

Note: The real-time decoding interfaces are experimental, and subject to change. Real-time decoding on Quantinuum's Helios-1 device is currently only available to partners and collaborators. Please email QCSupport@quantinuum.com for more information.

**Emulation vs. Hardware Modes:**

Emulation mode (``emulate=True``) is particularly valuable for testing the deployment setup without consuming hardware credits.

> However, the Quantinuum Helios-1 emulator does not support real-time decoder calls at this time.

Reviewer 2: Fixed in [92ebdf3](https://github.com/NVIDIA/cudaqx/pull/345/commits/92ebdf3c48c7be9bb53e9e02d7b1efd49672045a)
